### ISSUE TITLE
Feature Request: Add semver support

### ISSUE THREAD

**Issue by ciriarte** [carvel-dev/vendir#10]
## Context

**vendir** currently supports tracking specific refs and versions. I'd like to have support for specific version ranges:

```diff
  apiVersion: vendir.k14s.io/v1alpha1
  kind: Config
  directories:
  - path: vendor
    contents:
    - path: github.com/cloudfoundry/cf-k8s-networking
      git:
        url: https://github.com/cloudfoundry/cf-k8s-networking
-       ref: master
+       version: ^1.0.0
    includePaths:
    - cfroutesync/crds/**/*
    - install/ytt/networking/**/*
  - path: github.com/cloudfoundry/cc
    manual: {}
  - path: github.com/GoogleCloudPlatform/metacontroller
    manual: {}
  - path: local-dir
    directory:
      path: local-dir
```

So I can specify valid version ranges for each one of my dependencies.

**Comment by cppforlife** [comment]
i wonder if we should make this functionality more generic since vendir currently supports git and github release, but will also support sources such as docker images, helm charts, etc. ... or does it make sense to just keep it source specific?

**Comment by paulcwarren** [comment]
+1 for this.  

The syntax I have used for this in the past was something like this BTW `[1.0.0, 2.0.0)`; i.e. get the latest available from 1.0.0 up to but excluding 2.0.0.

**Comment by ciriarte** [comment]
> i wonder if we should make this functionality more generic since vendir currently supports git and github release, but will also support sources such as docker images, helm charts, etc. ... or does it make sense to just keep it source specific?

Yes. It shouldn't be specific to a source.

**Comment by ram-pivot** [comment]
+1 It will really helpful as we are also planning to consume other external components as libraries and this would help us to pin to a particular version.

**Comment by cppforlife** [comment]
sources:

- [ ] git (via tags)
- [ ] githubRelease (via tags)
- [ ] helm chart (via versions)
- [ ] image (via tags)
- [ ] http (???)

**Comment by anEXPer** [comment]
The Pivnet resource (and various other Concourse-ecosystem resources) accomplish some of this with regex filters; this allows expressive constraints of what versions will be included/excluded without having to create/document/maintain yet-another version specification DSL, or adopt one of the various (sometimes pretty complex) existing ones.

This also allows a distinction between the "explicit string" `version` key and a mutually-exclusive `version-regex` string, if desired.

Why not allow, for example,
`version-regex: ^v1\..*`

This also allows the use of sub-groups when using tags, and wanting to both be able to designate the "version" part while also being able to select from among multiple potentially version-matching tags, such as `release-v1.0.0` and `candidate-v1.0.0`, for example.

**Comment by cppforlife** [comment]
it seems to be pretty uncommon to use regex for version constraints compared to ">=1, <2" commonly found in various package managers.

**Comment by ciriarte** [comment]
I like https://github.com/Masterminds/semver because it tries to be compatible with other package managers (in version 3+). Also, they have a a cli to perform quick validations https://github.com/Masterminds/vert which is useful for scripts/pipeline composition.

e.g. I often end up mucking with `git describe --tags` but vert already supports the transformation/filtering: 

```bash
$ vert -g ^1 $(git describe --tags)
1.0.1+32.fef45
```

**Comment by cppforlife** [comment]
currently using hashicorps' go-version library (https://github.com/k14s/vendir/blob/develop/docs/versions.md).

**Comment by cppforlife** [comment]
@ciriarte first pass for git ref selection is on develop. example: https://github.com/k14s/vendir/blob/develop/examples/versionselection/vendir.yml. give it a go and lmk.

**Comment by ciriarte** [comment]
> @ciriarte first pass for git ref selection is on develop. example: https://github.com/k14s/vendir/blob/develop/examples/versionselection/vendir.yml. give it a go and lmk.

This is awesome. Will do and report back!

**Comment by cppforlife** [comment]
included in https://github.com/k14s/vendir/releases/tag/v0.11.0

**Comment by cppforlife** [comment]
switched to blang/semver in https://github.com/k14s/vendir/commit/cb32f5a5c820499f554029f7e6fe4867b5fbd289. this allows more control over when prerelease versions are included or not. added examples: https://github.com/k14s/vendir/blob/develop/docs/versions.md#examples. change will be available in v0.12.0+

**Comment by cppforlife** [comment]
updates are in https://github.com/k14s/vendir/releases/tag/v0.12.0
